File: Repositorio/engenharia-de-software-com-ia-aplicada-main/modulo08-arquitetura-de-sistemas-com-ia/modulo-01-fundamentos-ai-first/decision-framework-tool.js
/**
 * Ahirton Lopes · AI Architecture Toolkit
 * Artefato de Demo - Módulo 1.3
 *
 * Ferramenta: Checklist de Framework de Decisão de Três Perguntas, em código.
 * Referência: decision-framework-checklist.md (mesma pasta).
 *
 * Isso é lógica pura, 100% determinística: três perguntas booleanas viram uma
 * árvore de decisão simples, sem IA, sem modelo, sem chamada de rede nenhuma.
 * É o contraste pedagógico do módulo: quando a decisão é auditável e finita,
 * ela vira código com confiança total. Compare com o protótipo do Módulo 2.5
 * (react-agent-prototype.js), onde a decisão depende de julgamento de
 * linguagem natural e por isso não dá pra reduzir a uma árvore assim.
 *
 * Uso: node decision-framework-tool.js
 * Roda os testes automatizados e depois a demo, ambos no terminal.
 *
 * Mesmo padrão de nomenclatura dos outros protótipos da disciplina: este é o
 * par decision-framework-tool.js / decision_framework_tool.py. Esta versão em
 * JavaScript é a oficial da disciplina; a versão em Python é material de
 * referência espelhado, mesmo padrão dos outros módulos.
 */

'use strict';

const assert = require('assert').strict;

// As quatro classificações possíveis, exatamente como o texto do checklist.
// Centralizadas aqui pra evitar divergência de string entre a lógica e os testes.
const CLASSIFICACAO = {
  REGRA_DETERMINISTICA: 'Regra determinística',
  APPROVAL_GATE: 'Agente com Approval Gate obrigatório',
  AGENTE_AUTONOMO: 'Agente autônomo, com observabilidade completa',
  REGRA_ENUMERAVEL: 'Regra determinística (mesmo parecendo complexa, se é enumerável, é regra)',
};

/**
 * Aplica as três perguntas do checklist, em ordem, exatamente como a árvore:
 * Pergunta 1 decide sozinha quando é SIM (nem chega a olhar p2 ou p3); senão
 * passa pra Pergunta 2; senão passa pra Pergunta 3.
 *
 * @param {boolean} p1 Existe uma regra finita que cobre mais de 90% dos casos REAIS
 *                      (já observados, não hipotéticos)?
 * @param {boolean} p2 O erro é caro E a ação é irreversível (não dá pra desfazer depois)?
 * @param {boolean} p3 O comportamento da tarefa muda de acordo com o contexto de entrada?
 * @returns {string} uma das quatro classificações do checklist
 */
function classificarTarefa(p1, p2, p3) {
  if (p1) {
    return CLASSIFICACAO.REGRA_DETERMINISTICA;
  }
  if (p2) {
    return CLASSIFICACAO.APPROVAL_GATE;
  }
  if (p3) {
    return CLASSIFICACAO.AGENTE_AUTONOMO;
  }
  return CLASSIFICACAO.REGRA_ENUMERAVEL;
}

/**
 * Decompõe uma tarefa híbrida em subtarefas já classificadas, seguindo o
 * "Template: decompondo uma tarefa híbrida" do checklist. Não é uma lógica
 * nova: aplica classificarTarefa em cada subtarefa, uma de cada vez, e devolve
 * a mesma lista de entrada com o campo classificacao preenchido.
 *
 * @param {Array<{nome: string, tipo: string, p1: boolean, p2: boolean, p3: boolean}>} subtarefas
 * @returns {Array<{nome: string, tipo: string, p1: boolean, p2: boolean, p3: boolean, classificacao: string}>}
 */
function decomporTarefaHibrida(subtarefas) {
  return subtarefas.map((subtarefa) => ({
    ...subtarefa,
    classificacao: classificarTarefa(subtarefa.p1, subtarefa.p2, subtarefa.p3),
  }));
}

/* --------------------------------------------------------------------------
 * Testes automatizados (module assert nativo do Node, sem dependência externa)
 * Rodar com: node decision-framework-tool.js
 * -------------------------------------------------------------------------- */

let totalTestes = 0;
let testesComFalha = 0;

function testar(descricao, fn) {
  totalTestes += 1;
  try {
    fn();
    console.log(`  [OK] ${descricao}`);
  } catch (erro) {
    testesComFalha += 1;
    console.log(`  [FALHOU] ${descricao}`);
    console.log(`           ${erro.message}`);
  }
}

function rodarTestes() {
  console.log('== Testes: árvore pura (Bloco 1, as 4 combinações do checklist) ==');

  // Caso 1: p1=true -> Regra determinística, não importa p2/p3 ("qualquer").
  // Testamos as 4 combinações de p2/p3 pra provar que elas de fato não importam.
  testar('p1=true, p2=true,  p3=true  -> Regra determinística', () => {
    assert.equal(classificarTarefa(true, true, true), CLASSIFICACAO.REGRA_DETERMINISTICA);
  });
  testar('p1=true, p2=true,  p3=false -> Regra determinística', () => {
    assert.equal(classificarTarefa(true, true, false), CLASSIFICACAO.REGRA_DETERMINISTICA);
  });
  testar('p1=true, p2=false, p3=true  -> Regra determinística', () => {
    assert.equal(classificarTarefa(true, false, true), CLASSIFICACAO.REGRA_DETERMINISTICA);
  });
  testar('p1=true, p2=false, p3=false -> Regra determinística', () => {
    assert.equal(classificarTarefa(true, false, false), CLASSIFICACAO.REGRA_DETERMINISTICA);
  });

  // Caso 2: p1=false, p2=true -> Approval Gate, não importa p3 ("qualquer").
  testar('p1=false, p2=true, p3=true  -> Agente com Approval Gate obrigatório', () => {
    assert.equal(classificarTarefa(false, true, true), CLASSIFICACAO.APPROVAL_GATE);
  });
  testar('p1=false, p2=true, p3=false -> Agente com Approval Gate obrigatório', () => {
    assert.equal(classificarTarefa(false, true, false), CLASSIFICACAO.APPROVAL_GATE);
  });

  // Caso 3: p1=false, p2=false, p3=true -> Agente autônomo.
  testar('p1=false, p2=false, p3=true  -> Agente autônomo, com observabilidade completa', () => {
    assert.equal(classificarTarefa(false, false, true), CLASSIFICACAO.AGENTE_AUTONOMO);
  });

  // Caso 4: p1=false, p2=false, p3=false -> Regra determinística (enumerável).
  testar('p1=false, p2=false, p3=false -> Regra determinística (enumerável)', () => {
    assert.equal(classificarTarefa(false, false, false), CLASSIFICACAO.REGRA_ENUMERAVEL);
  });

  console.log();
  console.log('== Testes: decomposição de tarefa híbrida (Bloco 2, referência TrialForge) ==');

  // Nota honesta sobre a especificação: das quatro linhas da tabela de referência
  // do checklist (seção "Referência: TrialForge - Emenda de Protocolo"), só estas
  // duas abaixo mapeiam de forma limpa pra uma única resposta p1/p2/p3. "Rotear
  // pela criticidade" e "Regenerar documentos afetados" misturam regra e gate
  // condicional de um jeito mais sutil que a árvore pura de três perguntas não
  // representa sozinha, por isso ficam de fora do teste automatizado. Isso é uma
  // limitação real da árvore, não um bug desta implementação.
  const subtarefasReferencia = [
    {
      nome: 'Extrair o que mudou entre versões do protocolo',
      tipo: 'Extração/Interpretação',
      // Não é regra enumerável: é extração/interpretação de linguagem
      // natural sobre o texto do protocolo.
      p1: false,
      // A extração em si não causa, diretamente, um erro caro e irreversível.
      p2: false,
      // O comportamento muda bastante conforme o que de fato mudou entre
      // as duas versões do protocolo.
      p3: true,
    },
    {
      nome: 'Classificar o tipo de emenda (administrativa/substancial)',
      tipo: 'Decisão de Negócio',
      // Segue um critério fixo definido pela ANVISA, coberto no vídeo do
      // Módulo 1.3: regra finita cobre os casos reais.
      p1: true,
      p2: false,
      p3: false,
    },
  ];

  const resultado = decomporTarefaHibrida(subtarefasReferencia);

  testar('"Extrair o que mudou entre versões do protocolo" -> Agente autônomo, com observabilidade completa', () => {
    assert.equal(resultado[0].classificacao, CLASSIFICACAO.AGENTE_AUTONOMO);
  });
  testar('"Classificar o tipo de emenda (administrativa/substancial)" -> Regra determinística', () => {
    assert.equal(resultado[1].classificacao, CLASSIFICACAO.REGRA_DETERMINISTICA);
  });
  testar('decomporTarefaHibrida preserva nome e tipo de cada subtarefa na saída', () => {
    assert.equal(resultado[0].nome, subtarefasReferencia[0].nome);
    assert.equal(resultado[0].tipo, subtarefasReferencia[0].tipo);
    assert.equal(resultado[1].nome, subtarefasReferencia[1].nome);
    assert.equal(resultado[1].tipo, subtarefasReferencia[1].tipo);
  });

  console.log();
  console.log(
    `Total: ${totalTestes} teste(s), ${totalTestes - testesComFalha} passou(passaram), ` +
      `${testesComFalha} falhou(falharam).`
  );

  if (testesComFalha > 0) {
    throw new Error(`${testesComFalha} teste(s) falharam. A implementação não bate com a especificação.`);
  }
}

/**
 * Demo: imprime os 4 casos do Bloco 1 e a decomposição da tarefa híbrida do
 * Bloco 2 de forma legível no terminal, mesmo espírito da Missão Prática:
 * mostrar o mecanismo funcionando, não só os testes passando silenciosamente.
 */
function rodarDemo() {
  console.log();
  console.log('===== Demo: as 4 combinações da árvore pura =====');
  const casos = [
    { rotulo: 'Caso 1', p1: true, p2: false, p3: false },
    { rotulo: 'Caso 2', p1: false, p2: true, p3: false },
    { rotulo: 'Caso 3', p1: false, p2: false, p3: true },
    { rotulo: 'Caso 4', p1: false, p2: false, p3: false },
  ];
  for (const caso of casos) {
    const classificacao = classificarTarefa(caso.p1, caso.p2, caso.p3);
    console.log(`  ${caso.rotulo}: p1=${caso.p1}, p2=${caso.p2}, p3=${caso.p3} -> ${classificacao}`);
  }

  console.log();
  console.log('===== Demo: decompondo a tarefa híbrida (referência TrialForge, Emenda de Protocolo) =====');
  const subtarefas = [
    {
      nome: 'Extrair o que mudou entre versões do protocolo',
      tipo: 'Extração/Interpretação',
      p1: false,
      p2: false,
      p3: true,
    },
    {
      nome: 'Classificar o tipo de emenda (administrativa/substancial)',
      tipo: 'Decisão de Negócio',
      p1: true,
      p2: false,
      p3: false,
    },
  ];
  const decomposicao = decomporTarefaHibrida(subtarefas);
  for (const subtarefa of decomposicao) {
    console.log(`  [${subtarefa.tipo}] ${subtarefa.nome}`);
    console.log(`    p1=${subtarefa.p1}, p2=${subtarefa.p2}, p3=${subtarefa.p3} -> ${subtarefa.classificacao}`);
  }

  console.log();
  console.log('  Nota: as outras duas linhas da tabela de referência do checklist ("Rotear pela');
  console.log('  criticidade" e "Regenerar documentos afetados") misturam regra e gate condicional');
  console.log('  de um jeito que não mapeia limpo pra uma única resposta p1/p2/p3. Ficam de fora');
  console.log('  desta demo e dos testes automatizados por essa razão, não por engano.');
}

if (require.main === module) {
  rodarTestes();
  rodarDemo();
}

module.exports = { classificarTarefa, decomporTarefaHibrida, CLASSIFICACAO };

/*
 * Ahirton Lopes · AI Architecture Toolkit — UNIPDS: Arquitetura de Sistemas com IA
 * Prof. Ahirton Lopes, Ph.D. — GDE AI, Microsoft MVP, Senior Manager
 */

File: Repositorio/engenharia-de-software-com-ia-aplicada-main/modulo08-arquitetura-de-sistemas-com-ia/modulo-01-fundamentos-ai-first/decision_framework_tool.py
"""
Ahirton Lopes · AI Architecture Toolkit
Artefato de Demo - Módulo 1.3

Ferramenta: Framework de Decisão de Três Perguntas (checklist do Módulo 1.3).
Lógica pura, 100% determinística, sem IA: não chama nenhum modelo, não faz
nenhuma chamada de rede, não precisa de Ollama nem de chave de API alguma.
É a árvore de decisão do checklist (decision-framework-checklist.md, mesma
pasta) virando código, pergunta por pergunta.

Contraste pedagógico proposital com o protótipo do Módulo 2.5
(react-agent-prototype): lá a decisão depende de julgamento de linguagem
natural feito por um modelo, não é 100% determinística. Aqui a decisão é
uma árvore de três perguntas booleanas, então vira código com confiança
total, sem ambiguidade nenhuma. Esse contraste é o próprio valor
pedagógico deste artefato.

Versão oficial: JavaScript (decision-framework-tool.js), mesmo padrão de
nomenclatura kebab-case/snake_case já usado nos outros pares JS/Python da
disciplina. Esta versão em Python é material de referência espelhado, para
quem quiser comparar ou usar noutro contexto.

Requer: apenas biblioteca padrão do Python (módulos unittest e typing, já
embutidos). Nenhuma dependência externa, nenhuma chamada de rede.
Uso:    python3 decision_framework_tool.py
"""

import sys
import unittest
from typing import Any, Dict, List


def classificar_tarefa(p1: bool, p2: bool, p3: bool) -> str:
    """
    Aplica a árvore de três perguntas do checklist, na ordem exata do documento.

    P1 - Existe uma regra finita que cobre mais de 90% dos casos REAIS (já
         observados, não hipotéticos)?
    P2 - O erro é caro E a ação é irreversível (não dá pra desfazer depois)?
    P3 - O comportamento da tarefa muda de acordo com o contexto de entrada?

    A árvore tem curto-circuito, igual ao checklist original: se p1 é True,
    o "Fim da decisão" já aconteceu e p2/p3 nem chegam a ser avaliados (os
    parâmetros são recebidos, mas os valores deles não influenciam o
    resultado). O mesmo vale para p3 quando p1 é False e p2 é True.
    """
    if p1:
        return "Regra determinística"
    if p2:
        return "Agente com Approval Gate obrigatório"
    if p3:
        return "Agente autônomo, com observabilidade completa"
    return "Regra determinística (mesmo parecendo complexa, se é enumerável, é regra)"


def decompor_tarefa_hibrida(subtarefas: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Recebe uma lista de subtarefas, cada uma um dict com as chaves "nome",
    "tipo" (string livre, ex.: "Extração/Interpretação" ou "Decisão de
    Negócio"), "p1", "p2" e "p3" (as mesmas colunas do template "Decompondo
    uma tarefa híbrida" do checklist), e devolve uma lista nova, na mesma
    ordem, com uma chave a mais em cada item: "classificacao", o resultado de
    aplicar classificar_tarefa() naquela subtarefa especificamente.

    Não é uma lógica nova: é a mesma árvore de três perguntas, reaproveitada
    via classificar_tarefa(), aplicada item a item. A ideia central do
    checklist é justamente essa: decompor a tarefa "que parece uma coisa só"
    antes de classificar, porque uma tarefa híbrida raramente é 100% agente
    ou 100% regra de ponta a ponta.

    Escolha de projeto: a função não modifica os dicts de entrada (retorna
    dicts novos, com os campos originais copiados mais "classificacao"). Isso
    evita efeito colateral surpresa em quem chama a função passando a
    própria lista de referência do time.
    """
    resultado = []
    for subtarefa in subtarefas:
        classificacao = classificar_tarefa(subtarefa["p1"], subtarefa["p2"], subtarefa["p3"])
        resultado.append({**subtarefa, "classificacao": classificacao})
    return resultado


# ---------------------------------------------------------------------------
# Testes automatizados (unittest, biblioteca padrão do Python; sem instalar
# nada). Rodam sempre que o arquivo é executado diretamente, antes da demo.
# ---------------------------------------------------------------------------


class TestClassificarTarefa(unittest.TestCase):
    """Bloco 1: as 4 combinações da árvore pura, extraídas do checklist."""

    def test_p1_true_com_qualquer_p2_p3_e_regra_deterministica(self):
        for p2 in (True, False):
            for p3 in (True, False):
                with self.subTest(p2=p2, p3=p3):
                    self.assertEqual(classificar_tarefa(True, p2, p3), "Regra determinística")

    def test_p1_false_p2_true_com_qualquer_p3_e_approval_gate(self):
        for p3 in (True, False):
            with self.subTest(p3=p3):
                self.assertEqual(
                    classificar_tarefa(False, True, p3),
                    "Agente com Approval Gate obrigatório",
                )

    def test_p1_false_p2_false_p3_true_e_agente_autonomo(self):
        self.assertEqual(
            classificar_tarefa(False, False, True),
            "Agente autônomo, com observabilidade completa",
        )

    def test_p1_false_p2_false_p3_false_e_regra_deterministica_enumeravel(self):
        self.assertEqual(
            classificar_tarefa(False, False, False),
            "Regra determinística (mesmo parecendo complexa, se é enumerável, é regra)",
        )


class TestDecomporTarefaHibrida(unittest.TestCase):
    """
    Bloco 2: decomposição da tarefa híbrida de referência do checklist
    (seção "Referência: TrialForge - Emenda de Protocolo").

    Usa só as duas subtarefas da tabela de referência que mapeiam limpo pra
    uma única resposta p1/p2/p3: "Extrair o que mudou entre versões do
    protocolo" e "Classificar o tipo de emenda (administrativa/substancial)".

    As outras duas linhas da tabela do checklist ("Rotear pela criticidade"
    e "Regenerar documentos afetados") misturam regra e gate condicional de
    um jeito mais sutil, que não mapeia limpo pra uma única resposta
    p1/p2/p3 sem inventar valores só pra forçar um teste a passar. Isso é
    uma limitação real e honesta da árvore pura de três perguntas (ela
    classifica o componente predominante de uma subtarefa já bem recortada,
    não decide composições internas de regra+gate dentro da mesma
    subtarefa), não um bug desta implementação.
    """

    def test_decompoe_as_duas_subtarefas_de_referencia_do_checklist(self):
        subtarefas = [
            {
                "nome": "Extrair o que mudou entre versões do protocolo",
                "tipo": "Extração/Interpretação",
                # Não é regra enumerável: é extração/interpretação de
                # linguagem natural sobre o texto do protocolo.
                "p1": False,
                # A extração em si não causa, diretamente, um erro caro e
                # irreversível.
                "p2": False,
                # O comportamento muda bastante conforme o que de fato
                # mudou entre as duas versões do protocolo.
                "p3": True,
            },
            {
                "nome": "Classificar o tipo de emenda (administrativa/substancial)",
                "tipo": "Decisão de Negócio",
                # Segue um critério fixo definido pela ANVISA, coberto no
                # vídeo do Módulo 1.3: regra finita cobre os casos reais.
                "p1": True,
                "p2": False,
                "p3": False,
            },
        ]

        resultado = decompor_tarefa_hibrida(subtarefas)

        self.assertEqual(
            resultado[0]["classificacao"], "Agente autônomo, com observabilidade completa"
        )
        self.assertEqual(resultado[1]["classificacao"], "Regra determinística")

        # A decomposição preserva os campos originais (nome, tipo, p1..p3)
        # e só acrescenta "classificacao"; não descarta nem reordena nada, e
        # não modifica a lista/dicts de entrada (ver docstring da função).
        self.assertEqual(len(resultado), len(subtarefas))
        self.assertEqual(resultado[0]["nome"], subtarefas[0]["nome"])
        self.assertEqual(resultado[1]["tipo"], subtarefas[1]["tipo"])
        self.assertIsNot(resultado[0], subtarefas[0])
        self.assertNotIn("classificacao", subtarefas[0])


# ---------------------------------------------------------------------------
# Demo (roda quando o arquivo é executado diretamente, depois dos testes)
# ---------------------------------------------------------------------------


def executar_demo() -> None:
    """
    Demo legível no terminal: os 4 casos do Bloco 1 e a decomposição da
    tarefa híbrida de referência do Bloco 2. Mesmo espírito da Missão
    Prática: mostrar o mecanismo funcionando, não só os testes passando
    silenciosamente.
    """
    print("=" * 72)
    print("Bloco 1: as 4 combinações da árvore pura de três perguntas")
    print("=" * 72)

    casos_bloco_1 = [
        ("P1=True,  P2=False, P3=False", True, False, False),
        ("P1=False, P2=True,  P3=False", False, True, False),
        ("P1=False, P2=False, P3=True ", False, False, True),
        ("P1=False, P2=False, P3=False", False, False, False),
    ]
    for rotulo, p1, p2, p3 in casos_bloco_1:
        print(f"  {rotulo} -> {classificar_tarefa(p1, p2, p3)}")

    print()
    print("=" * 72)
    print("Bloco 2: decompondo a tarefa híbrida de referência")
    print("(TrialForge - Emenda de Protocolo)")
    print("=" * 72)

    subtarefas = [
        {
            "nome": "Extrair o que mudou entre versões do protocolo",
            "tipo": "Extração/Interpretação",
            "p1": False,
            "p2": False,
            "p3": True,
        },
        {
            "nome": "Classificar o tipo de emenda (administrativa/substancial)",
            "tipo": "Decisão de Negócio",
            "p1": True,
            "p2": False,
            "p3": False,
        },
    ]
    for subtarefa in decompor_tarefa_hibrida(subtarefas):
        print(f"  Subtarefa: {subtarefa['nome']}")
        print(f"    Tipo: {subtarefa['tipo']}")
        print(f"    P1={subtarefa['p1']}, P2={subtarefa['p2']}, P3={subtarefa['p3']}")
        print(f"    Classificação: {subtarefa['classificacao']}")
        print()

    print(
        "  Nota: as outras duas linhas da tabela de referência do checklist\n"
        "  ('Rotear pela criticidade' e 'Regenerar documentos afetados') não\n"
        "  entram nesta demo de propósito, pelo mesmo motivo documentado em\n"
        "  TestDecomporTarefaHibrida: misturam regra e gate condicional de um\n"
        "  jeito que não mapeia limpo pra uma única resposta P1/P2/P3."
    )


if __name__ == "__main__":
    suite = unittest.defaultTestLoader.loadTestsFromModule(sys.modules[__name__])
    resultado_testes = unittest.TextTestRunner(verbosity=2).run(suite)

    if not resultado_testes.wasSuccessful():
        print("\n[Testes falharam] Corrigindo a lógica antes de rodar a demo.")
        sys.exit(1)

    print()
    executar_demo()

# Ahirton Lopes · AI Architecture Toolkit — UNIPDS: Arquitetura de Sistemas com IA
# Prof. Ahirton Lopes, Ph.D. — GDE AI, Microsoft MVP, Senior Manager
